---
title: "manuscript_public_code"
output: html_document
date: "2025-04-14"
---

This is an R-markdown to completely re-generate the figures from the ReVSeq Manuscript.

```{r}
#sourcing all packages
```


```{r}
#sourcing all files 

#must upload virus abbreviation CSV

#overview 
source("fig_1_data_format.R")
source("fig_1_plot.R")

#PCR to seq comparison
source("fig_2_data_format.R")
source("fig_2_plot.R")

#co-infections
source("fig_3_data_format.R")
source("fig_3_plot.R")

#coverage plots
source("coverage_genome_data_format.R") #requires external docs > to add (bedfile and gff files)
source("coverage_genome_plot.R")

```

```{r}
hq_data<-read.csv("data/final_analysis/1_19_05_25_pull/manuscript_public/sample_data.csv")
pcr_data <- read.csv("data/final_analysis/1_19_05_25_pull/manuscript_public/pcr_data.csv")
detected_data <- read.csv("data/final_analysis/1_19_05_25_pull/manuscript_public/sample_data_detected.csv")
sentinella_data <- get_sentinella_data()

hq_data_old<-read.csv("data/final_analysis/1_manuscript_public/sample_data.csv")
pcr_data_old <- read.csv("data/final_analysis/1_manuscript_public/pcr_data.csv")
detected_data_old <- read.csv("data/final_analysis/1_manuscript_public/sample_data_detected.csv")
#sentinella_data <- get_sentinella_data()
```



# Figure 1: seasonal overview

```{r}
substrains_to_highlight = ""
df_plt <- format_grid_pathogen_plot(hq_data, pcr=FALSE)
freq_data<-format_frequencies_plot(pcr_data, hq_data, detected_data ,substrain_to_highlight = substrains_to_highlight)

#substrain_to highlight must be a vector
make_figure_one(df_plt, freq_data, sentinella_data ,highlight = FALSE, save=FALSE, detected=FALSE )


```
# Figure 2: PCR / sequencing comparison

```{r}
#OLD
#all strains included in the PCR panel
pcr_panel_strains <- (pcr_data_old %>%  separate_rows(strains_PCR, panels_PCR, values_PCR, sep = ",\\s*") )$strains_PCR %>% unique()

formatted_pcr_to_detected<-format_pcr_to_detect_data(pcr_data_old, detected_data_old, pcr_panel_strains)
formatted_detected_to_hq<-format_detect_to_hq_data(detected_data_old, hq_data_old, pcr_panel_strains)

make_figure_comparison(formatted_pcr_to_detected, formatted_detected_to_hq, save=FALSE)
```



```{r}
#all strains included in the PCR panel
pcr_panel_strains <- (pcr_data %>%  separate_rows(strains_PCR, panels_PCR, values_PCR, sep = ",\\s*") )$strains_PCR %>% unique()

formatted_pcr_to_detected<-format_pcr_to_detect_data(pcr_data, detected_data, pcr_panel_strains)
formatted_detected_to_hq<-format_detect_to_hq_data(detected_data, hq_data, pcr_panel_strains)

make_figure_comparison(formatted_pcr_to_detected, formatted_detected_to_hq, save=FALSE)

```

# Figure 3: Co-infections in high-quality sequences

```{r}
matrix_co <- format_matrix_co_infection_plot(hq_data)
percent_co <- format_percent_count_co_infection_plot(hq_data)

make_figure_3(matrix_co, percent_co, save=FALSE)
```
# Figure 4: Whole-genome coverage

```{r}
# RSV-A
#get data
virus_strain <- "Respiratory syncytial virus (type A)"
rsv_a_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(rsv_a_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

# Parainfluenza-3
virus_strain <- "Human parainfluenza virus 3"
hpiv_3_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hpiv_3_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```
#hmpv test
```{r}
# HMPV
virus_strain <- "Human metapneumovirus A"
hpiv_3_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hpiv_3_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)
```


# Figure 5: Phylogenetic trees

```{r}

```

# Supp table counts 
 
```{r}
df_pcr <-pcr_data %>% separate_longer_delim(cols = c(strains_PCR, panels_PCR, values_PCR), delim = ",") %>%
    # Remove leading/trailing whitespace
    mutate(strains_PCR = str_trim(strains_PCR),
           panels_PCR = str_trim(panels_PCR),
           values_PCR= str_trim(values_PCR)) %>% 
  mutate(
    panel_type = ifelse(grepl("resp", panels_PCR), "resp", ifelse(panels_PCR == "flua" | panels_PCR == "flub" | panels_PCR == "rspc", "p2", "p1"))
      ) %>% 
  filter(!is.na(strains_PCR)) %>% 
  dplyr::count(panel_type, strains_PCR)

df_detect <- detected_data %>% dplyr::count(strain_name)
df_hq <-hq_data %>% dplyr::count(strain_name)


```

# Supp wastewater 


```{r}
sentinella_df <- get_sentinella_data() 
wastewater_df <- get_wastewater_data()
pathogens <- sentinella_df$pathogen_name %>% unique() %>% unlist()

pcr_dat_supp <- pcr_data %>% 
  separate_longer_delim(cols = c(strains_PCR, panels_PCR, values_PCR), delim = ",") %>%
    # Remove leading/trailing whitespace
  mutate(strains_PCR = str_trim(strains_PCR),
           panels_PCR = str_trim(panels_PCR),
           values_PCR= str_trim(values_PCR)) %>% 
  rename(substrain_name = strains_PCR)


pcr_dat<-format_grid_pathogen_plot(pcr_dat_supp, pcr=TRUE) %>% drop_na()
detected_dat <- format_grid_pathogen_plot(detected_data, pcr=FALSE) %>% drop_na()
sequencing_data <- format_grid_pathogen_plot(hq_data, pcr=FALSE) %>% drop_na()


make_supplementary_figure_2(sentinella_df, pcr_dat, sequencing_data, detected_dat, wastewater_df , pathogens=pathogens ,save=FALSE)
```

# Supp Coverage plots

```{r}
# RSV-B
#get data
virus_strain <- "Human Respiratory syncytial virus 9320 (type B)"
rsv_b_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(rsv_b_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)


```



```{r}
#H1N1
#get data
virus_strain <- "2015(H1N1)"
h1n1_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(h1n1_data, virus_strain)
figs<-make_figure_coverage_genome(df, virus_strain, save=FALSE)


plots <- list( figs$h1n1_HA, figs$h1n1_NA)

plot_all<-plot_grid(plotlist = plots,
          #nrow = length(virus_strains), 
          ncol = 2,
          rel_heights = c(1,1),
          align = "v",
          axis="l")

plot_all
#ggsave("images/final_analysis/1_19_05_25_pull/coverage_plots/h1n1_HA_NA.pdf",plot_all , dpi="retina", units="mm", width=174, height=123)
```


# sampling even throughout the season?

```{r}
df_sampling <-pcr_data %>% 
  mutate(date = as.Date(ent_date),
         month_year_label = format(date, "%b %Y"),
    # Convert to factor with levels in chronological order
    month_year_label = factor(month_year_label, levels = unique(month_year_label[order(date)])),
         panel_used = case_when(
           grepl("resp", panels_PCR) ~ "RESP",
           grepl("wuhan0", panels_PCR) | grepl("4p", panels_PCR) ~ "SARS/INF/RSV",
           grepl("rspc", panels_PCR) ~ "INF/RSV",
           is.na(panels_PCR) ~ "RESP",
           TRUE ~"INF/RSV"
         )
  ) %>% 
  select(pseudonymized_id, date, month_year_label, panel_used)

  
sampling_rate<-ggplot(data=df_sampling %>% dplyr::count(month_year_label, panel_used)) +
  geom_bar(aes(x=month_year_label, y= n), stat = "identity") + facet_wrap(~panel_used) +
  scale_x_discrete(drop = FALSE) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

sampling_rate

ggsave("images/final_analysis/1_19_05_25_pull/sampling_rate.pdf",sampling_rate , dpi="retina", units="mm", width=174, height=123)

```




