---
title: "Coverage plots for all single-stranded viruses"
output: github_document
date: "2025-07-02"
---

```{r}
#sourcing all packages
library(dplyr)
library(stringr)
library(purrr)
library(ggplot2)
library(cowplot)
library(tidyr)
library(ggpubr)
```

```{r}
#sourcing functions for plotting

#coverage plots
source("Sequencing_results/coverage_genome_data_format.R") #requires external docs > to add (bedfile and gff files)
source("Sequencing_results/coverage_genome_plot.R")

```

```{r}
#data set needed
hq_data<-read.csv("Data/data/hq_data.csv")
```

# RSV-A 

```{r, warning=FALSE}
# RSV-A
#get data
virus_strain <- "Respiratory syncytial virus (type A)"
rsv_a_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(rsv_a_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# RSV-B

```{r, warning=FALSE}
# RSV-B
#get data
virus_strain <- "Human Respiratory syncytial virus 9320 (type B)"
rsv_b_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(rsv_b_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)
```

# HPIV-3

```{r, warning=FALSE}
# Parainfluenza-3
virus_strain <- "Human parainfluenza virus 3"
hpiv_3_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hpiv_3_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# HPIV-2

```{r, warning=FALSE}
# Parainfluenza-2
virus_strain <- "Human parainfluenza virus 2"
hpiv_2_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hpiv_2_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# HPIV-1

```{r, warning=FALSE}
# Parainfluenza-1
virus_strain <- "Human parainfluenza virus 1"
hpiv_1_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hpiv_1_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

#SARS-CoV-2

```{r, warning=FALSE}
virus_strain <- "SARS-CoV-2"
sars_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(sars_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)
```

# HMPV

```{r, warning=FALSE}
# HMPV
#get data
virus_strain <- "metapneumovirus" 
hmpv_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hmpv_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)
```


# A/H1N1 HA/NA segments

```{r, warning=FALSE}
#H1N1
#get data
virus_strain <- "2015(H1N1)"
h1n1_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(h1n1_data, virus_strain)
figs<-make_figure_coverage_genome(df, virus_strain, save=FALSE)


plots <- list( figs$h1n1_HA, figs$h1n1_NA)

plot_all<-plot_grid(plotlist = plots,
          #nrow = length(virus_strains), 
          ncol = 2,
          rel_heights = c(1,1),
          align = "v",
          axis="l")

plot_all
```

# A/H3N2 HA/NA segments

```{r, warning=FALSE}
#H3N2
#get data
virus_strain <- "2012(H3N2)"
h3n2_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(h3n2_data, virus_strain)
figs<-make_figure_coverage_genome(df, virus_strain, save=FALSE)


plots <- list( figs$h3n2_HA, figs$h3n2_NA)

plot_all<-plot_grid(plotlist = plots,
          #nrow = length(virus_strains), 
          ncol = 2,
          rel_heights = c(1,1),
          align = "v",
          axis="l")

plot_all
```

# B/Victoria

```{r, warning=FALSE}
#B/Victoria
#get data
virus_strain <- "B/Brisbane"
vic_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(vic_data, virus_strain)
figs<-make_figure_coverage_genome(df, virus_strain, save=FALSE)


plots <- list( figs$vic_HA, figs$vic_NA)

plot_all<-plot_grid(plotlist = plots,
          #nrow = length(virus_strains), 
          ncol = 2,
          rel_heights = c(1,1),
          align = "v",
          axis="l")

plot_all
```

# HKU1

```{r, warning=FALSE}
# seasonal coronavirus HKU1
virus_strain <- "HKU1"
hku1_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(hku1_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# OC43

```{r, warning=FALSE}
# seasonal coronavirus OC43
virus_strain <- "OC43"
oc43_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(oc43_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# NL63

```{r, warning=FALSE}
# seasonal coronavirus NL63
virus_strain <- "NL63"
nl63_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(nl63_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```

# 229E

```{r, warning=FALSE}
# seasonal coronavirus 229E
virus_strain <- "229E"
e229_data <- hq_data %>% filter(grepl(virus_strain, substrain_name, fixed=TRUE))

df <- format_coverage_plot_data(e229_data, virus_strain)
make_figure_coverage_genome(df, virus_strain, save=FALSE)

```